### PR TITLE
[MultiKueue] fix using MultiKueue when some integrations are disabled

### LINKED ISSUE #3582: [Multi Kueue] Client Connection Failures with removed integrations
**What happened**:
I set up Multi Kueue as per the [documentation](https://kueue.sigs.k8s.io/docs/tasks/manage/setup_multikueue/).
I removed `"jobset.x-k8s.io/jobset"` from the list of integrations in the kueue-manager-config Config Map as I did not intend on using JobSets. This was included in the application of the Kueue manifests.

After setting up my Multi Kueue environment I ran the example test commands to which I got the following errors:
```
CQ - Active: False Reason: AdmissionCheckInactive Message: Can't admit new workloads: references inactive AdmissionCheck(s): [sample-multikueue].
AC - Active: False Reason: NoUsableClusters Message: Inactive clusters: [multikueue-test-worker1]
MC - Active: False Reason: ClientConnectionFailed Message: no matches for kind "JobSet" in version "jobset.x-k8s.io/v1alpha2"
```

**What you expected to happen**:
I expected that when an integration is disabled in the Kueue manager config it should be reflected in MultiKueue meaning no additional CRDs should have to be installed for Job types the user has no intention on using.

**How to reproduce it (as minimally and precisely as possible)**:
* Setup your Clusters for Multi Kueue. (Ensure JobSet is removed from the list of integrations in the manager Config Map)
* Do not install the JobSet Controller/CRDs
* Remove JobSet Cluster Roles for the multikueue-sa
* Examine your Multi Kueue Cluster CR for the failure message.
 
**Anything else we need to know?**:
This same behaviour would happen if it was the MPI Operator that was not installed. 
**Environment**:
- Kubernetes version (use `kubectl version`): v1.28
- OpenShift version: 4.15.37
- Kueue version (use `git describe --tags --dirty --always`): v0.9.0
- Cloud provider or hardware configuration: AWS
- Install tools: kubectl
- Others:
- MPI Operator: v0.6.0
- KubeFlow Training Operator: v1.8.1

### PR DESCRIPTION
**Pull request by mszadkow**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
MK adapter didn't contain the condition to only allow enabled integrations.
Thus while using multi cluster setup, all registered MK adapters were used to setup controllers.
In result we get `ClientConnectionFailed error with no matches for kind "XYZ" in "version v1" message`

#### Which issue(s) this PR fixes:

Fixes #3582

#### Does this PR introduce a user-facing change?

```release-note
Fix the bug which prevented the use of MultiKueue if there is a CRD which is not installed
and removed from the list of enabled integrations.
```

### CODE REVIEW

**Comment by mimowo** [comment]
/assign @mbobrovskyi 
for the first pass

File: test/integration/multikueue/suite_test.go [hunk 1]
```diff
@@ -213,10 +214,17 @@ func managerAndMultiKueueSetup(ctx context.Context, mgr manager.Manager, gcInter
 	err := multikueue.SetupIndexer(ctx, mgr.GetFieldIndexer(), managersConfigNamespace.Name)
 	gomega.Expect(err).NotTo(gomega.HaveOccurred())
 
+	adapters, err := jobframework.GetMultiKueueAdapters(sets.New([]string{
+		"batch/job", "kubeflow.org/mpijob", "ray.io/rayjob", "ray.io/raycluster",
+		"jobset.x-k8s.io/jobset", "kubeflow.org/mxjob", "kubeflow.org/paddlejob",
+		"kubeflow.org/pytorchjob", "kubeflow.org/tfjob", "kubeflow.org/xgboostjob"}...))
```
Comment (trasc) [inline test/integration/multikueue/suite_test.go:220]:
```suggestion
	adapters, err := jobframework.GetMultiKueueAdapters(sets.New[string](
		"batch/job", "kubeflow.org/mpijob", "ray.io/rayjob", "ray.io/raycluster",
		"jobset.x-k8s.io/jobset", "kubeflow.org/mxjob", "kubeflow.org/paddlejob",
		"kubeflow.org/pytorchjob", "kubeflow.org/tfjob", "kubeflow.org/xgboostjob"))
```

Also the slice of enabled integrations  can be an argument  to this function,  whit this you can move the tests from "./failedsetup" , in this suite (but sure in their own `var _ = ginkgo.Describe(` )

File: cmd/kueue/main.go [hunk 2]
```diff
@@ -264,10 +265,16 @@ func setupControllers(ctx context.Context, mgr ctrl.Manager, cCache *cache.Cache
 	}
 
 	if features.Enabled(features.MultiKueue) {
+		adapters, err := jobframework.GetMultiKueueAdapters(sets.New(cfg.Integrations.Frameworks...))
+		if err != nil {
+			setupLog.Error(err, "Could not setup MultiKueue controller")
```
Comment (trasc) [inline cmd/kueue/main.go:270]:
```suggestion
			setupLog.Error(err, "Could not get the enabled multikueue adapters")
```

File: pkg/controller/admissionchecks/multikueue/multikueuecluster_test.go [hunk 3]
```diff
@@ -376,7 +377,8 @@ func TestUpdateConfig(t *testing.T) {
 			builder = builder.WithStatusSubresource(slices.Map(tc.clusters, func(c *kueue.MultiKueueCluster) client.Object { return c })...)
 			c := builder.Build()
 
-			adapters, _ := jobframework.GetMultiKueueAdapters()
+			enabledIntegrations := sets.New([]string{"batch/job"}...)
+			adapters, _ := jobframework.GetMultiKueueAdapters(enabledIntegrations)
```
Comment (trasc) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster_test.go:381]:
```suggestion
			adapters, _ := jobframework.GetMultiKueueAdapters(sets.New("batch/job"))
```
The same goes in a couple of other places.

File: pkg/controller/admissionchecks/multikueue/workload_test.go [hunk 4]
```diff
@@ -51,6 +52,7 @@ var (
 )
 
 func TestWlReconcile(t *testing.T) {
+	t.Cleanup(jobframework.EnableIntegrationsForTest(t, "batch/job"))
```
Comment (trasc) [inline pkg/controller/admissionchecks/multikueue/workload_test.go:55]: Is this still needed?

File: test/integration/multikueue/failedsetup/multikueue_test.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+		gomega.Expect(util.DeleteNamespace(worker1TestCluster.ctx, worker1TestCluster.client, worker1Ns)).To(gomega.Succeed())
+		util.ExpectObjectToBeDeleted(managerTestCluster.ctx, managerTestCluster.client, managerCq, true)
+		util.ExpectObjectToBeDeleted(worker1TestCluster.ctx, worker1TestCluster.client, worker1Cq, true)
+		util.ExpectObjectToBeDeleted(managerTestCluster.ctx, managerTestCluster.client, multikueueAC, true)
+		util.ExpectObjectToBeDeleted(managerTestCluster.ctx, managerTestCluster.client, managerMultiKueueConfig, true)
+		util.ExpectObjectToBeDeleted(managerTestCluster.ctx, managerTestCluster.client, workerCluster1, true)
+		util.ExpectObjectToBeDeleted(managerTestCluster.ctx, managerTestCluster.client, managerMultikueueSecret1, true)
+	})
+
+	ginkgo.It("Should not create a MPIJob workload, because MPIJob CRD is not installed", func() {
```
Comment (trasc) [inline test/integration/multikueue/failedsetup/multikueue_test.go:143]:
```suggestion
	ginkgo.It("Should not create a MPIJob workload, when MPIJob adapter  is not enabled", func() {
```

File: test/integration/multikueue/failedsetup/multikueue_test.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+				},
+			).
+			Obj()
+
+		err := managerTestCluster.client.Create(managerTestCluster.ctx, mpijob)
+		gomega.Expect(err).Should(gomega.HaveOccurred())
+		gomega.Expect(apimeta.IsNoMatchError(err)).To(gomega.BeTrue())
+
+		wlLookupKey := types.NamespacedName{Name: workloadmpijob.GetWorkloadNameForMPIJob(mpijob.Name, mpijob.UID), Namespace: managerNs.Name}
+		ginkgo.By("workload not found", func() {
```
Comment (trasc) [inline test/integration/multikueue/failedsetup/multikueue_test.go:180]: The MpiJob integration should be enabled, and we should check that it's workload get's rejected by multikueue die to missing adapter.

Comment (mszadkow) [reply test/integration/multikueue/failedsetup/multikueue_test.go:180]: This is another test imho, equally important

File: pkg/controller/admissionchecks/multikueue/workload_test.go (hunk 4 above)
Comment (mszadkow) [reply pkg/controller/admissionchecks/multikueue/workload_test.go:55]: not needed indeed

File: test/integration/multikueue/failedsetup/multikueue_test.go (hunk 6 above)
Comment (trasc) [reply test/integration/multikueue/failedsetup/multikueue_test.go:180]: The workload associated to an MPIJob is created by the MPIJob integration, agnostic to multikueue and multikueue configuration.

File: test/integration/multikueue/multikueue_test.go [hunk 7]
```diff
@@ -1638,6 +1644,157 @@ var _ = ginkgo.Describe("Multikueue no GC", ginkgo.Ordered, ginkgo.ContinueOnFai
...
+		gomega.Expect(util.DeleteNamespace(worker1TestCluster.ctx, worker1TestCluster.client, worker1Ns)).To(gomega.Succeed())
+		util.ExpectObjectToBeDeleted(managerTestCluster.ctx, managerTestCluster.client, managerCq, true)
+		util.ExpectObjectToBeDeleted(worker1TestCluster.ctx, worker1TestCluster.client, worker1Cq, true)
+		util.ExpectObjectToBeDeleted(managerTestCluster.ctx, managerTestCluster.client, multikueueAC, true)
+		util.ExpectObjectToBeDeleted(managerTestCluster.ctx, managerTestCluster.client, managerMultiKueueConfig, true)
+		util.ExpectObjectToBeDeleted(managerTestCluster.ctx, managerTestCluster.client, workerCluster1, true)
+		util.ExpectObjectToBeDeleted(managerTestCluster.ctx, managerTestCluster.client, managerMultikueueSecret1, true)
+	})
+
+	ginkgo.It("Should not create a MPIJob workload, when MPIJob adapter is not enabled", func() {
```
Comment (mimowo) [inline test/integration/multikueue/multikueue_test.go:1750]: This test looks ok, but I think we need to add a test case for using the positive scenario (which was broken), that some integrations can be used (e.g. batch/Job) while not all are enabled, maybe "should create batch/Job workload when some frameworks are disabled"

File: test/integration/multikueue/multikueue_test.go [hunk 8]
```diff
@@ -1638,6 +1644,157 @@ var _ = ginkgo.Describe("Multikueue no GC", ginkgo.Ordered, ginkgo.ContinueOnFai
 	})
 })
 
+var _ = ginkgo.Describe("Multikueue with disabled integrations", ginkgo.Ordered, ginkgo.ContinueOnFailure, func() {
```
Comment (mimowo) [inline test/integration/multikueue/multikueue_test.go:1647]:
```suggestion
var _ = ginkgo.Describe("Multikueue with some integration disabled", ginkgo.Ordered, ginkgo.ContinueOnFailure, func() {
```

**Review by mimowo — COMMENTED** [review]
overall lgtm, but I would like yet to add the "positive" integration test to clearly demonstrate the issue is fixed

**Comment by mimowo** [comment]
/release-note-edit
```release-note
Fix the bug which prevented the use of MultiKueue if there is a CRD which is not installed
and removed from the list of enabled integrations.
```
/retitle [MultiKueue] fix using MultiKueue when some integrations are disabled

File: test/integration/multikueue/multikueue_test.go [hunk 9]
```diff
@@ -1638,6 +1644,246 @@ var _ = ginkgo.Describe("Multikueue no GC", ginkgo.Ordered, ginkgo.ContinueOnFai
 	})
 })
 
+var _ = ginkgo.Describe("Multikueue with some integration disabled", ginkgo.Ordered, ginkgo.ContinueOnFailure, func() {
```
Comment (mimowo) [inline test/integration/multikueue/multikueue_test.go:1647]:
```suggestion
var _ = ginkgo.Describe("Multikueue when not all integrations are enabled", ginkgo.Ordered, ginkgo.ContinueOnFailure, func() {
```
nit

**Pull request merged**